cmd/dex/config.go

```diff
@@ -129,6 +129,10 @@ func (p *password) UnmarshalJSON(b []byte) error {
 
 // OAuth2 describes enabled OAuth2 extensions.
 type OAuth2 struct {
+	// list of allowed grant types,
+	// defaults to all supported types
+	GrantTypes []string `json:"grantTypes"`
```

Comment:
There is an exciting nug when the grantTypes contains the `password` grant.

The discovery endpoint will return 
```
"grant_types_supported": [
  "password",
  "password"
],
```

For unknown grant type, it will return this claim. For example, for `test` grantTypes it will be 
```
"grant_types_supported": [
  "test"
],
```
which is not valid.

I think grant types should be applied as the final filter here only to find the intersection between known enabled grants and allowed grants from the config.
https://github.com/dexidp/dex/blob/47a0e067a174119224b49aba1549eb8af1550112/server/server.go#L216-L237

Reply:
I've implemented this